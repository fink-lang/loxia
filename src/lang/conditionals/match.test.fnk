{fink2js} = import '../../testing'
{describe, it, only, expect, to_match_snapshot} = import '@fink/jest'


describe 'match', fn:
  it 'compiles', fn:
    expect
      fink2js`
        match shrub:
          # simple value
          123:
            # result
            spam

          # object matching
          {foo: 4, ni}:
            spam + ni

          # array matching
          [1, 2, ni]:
            ni

          # fallback if none of above match
          else:
            shrub
      `
      to_match_snapshot


  it 'compiles without else', fn:
    expect
      fink2js`
        match shrub:
          {foo: 4, ni}: spam + ni
          123: spam
      `
      to_match_snapshot


  it 'compiles nested pattern', fn:
    expect
      fink2js`
        match shrub:
          # deep obj comparison
          {foo: {}}: spam + ni
          {foo: []}: spam + ni
          {foo: 4, ni: {na, nu}}: spam + ni
          {xx: 1, foo: {bar: 'spam'}, shrub: {na: 'nu'}}: ni
          {ni: {len: 1}, na: {len: 1}}: na
          # calculated props
          {(a): {b, c}}: ni
          # spread
          {(foo): {}, ...{bar, spam}}: ni
      `
      to_match_snapshot

    expect
      fink2js`
        match shrub:
          # deep array comparison
          [1, 2, [ni]]: ni
          [1, 2, [], {}]: foo
          [1, [2, 3], 4]: bar
          [1, [2, [3, 4]], [5, 6], 7]: spam
      `
      to_match_snapshot

    expect
      fink2js`
        # spread and ignored elems
        match shrub:
          [1, ,2]: foo
          [1, ...{length: 0}]: foo
          [1, ..., 3]: foo
          [1, ...{length: 0}, 3]: foo
      `
      to_match_snapshot


  it 'compiles nested match', fn:
    expect
      fink2js`
        test = fn:
          match foo:
            bar:
              match spam:
                shrub: ni
              bar
      `
      to_match_snapshot


  it 'compiles patial application', fn:
    expect
      fink2js`
        match {spam: 123}:
          # simple unary
          !?: ni

          # simple binary
          ? > 123: ni

          # simple call
          ?(): ni

          # simple as call arg
          shrub ?: ni


          # nested unary
          {spam: !?}: ni

          # nested binary
          {spam: ? > 123}: ni

          # nested call
          {spam: ?()}: ni

          # nested member call
          {spam: ?.is_foo()}: ni

          # nested as call arg
          {spam: is_foo ?}: ni


          # nested unary
          [!?]: ni

          # nested binary
          [? > 123]: ni

          # nested call
          [?()]: ni

          # nested member call
          [?.is_foo()]: ni

          # nested as call arg
          [is_foo ?]: ni
      `
      to_match_snapshot


  it 'compiles do-expr', fn:
    expect
      fink2js`
        foo = match bar:
          spam:
            ni
            throw foobar
          else:
            shrub
            throw ni
      `
      to_match_snapshot

