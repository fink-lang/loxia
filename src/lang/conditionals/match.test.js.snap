// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compiles logical 1`] = `
"{
  ˆmatch_2: {
    const ˆvalue_1 = shrub;

    if (ˆvalue_1 !== undefined && ˆvalue_1 !== null) {
      const {
        \\"foo\\": ˆp_3,
        \\"ni\\": ˆp_4
      } = ˆvalue_1;

      if (ˆp_3 === 4 && ˆp_4 === ni) {
        spam + ni;
        break ˆmatch_2;
      }
    }

    if (ˆvalue_1 === 123) {
      spam;
      break ˆmatch_2;
    }

    if (ˆvalue_1 !== undefined && ˆvalue_1 !== null && typeof ˆvalue_1[Symbol.iterator] === \\"function\\") {
      const [ˆa_5, ˆa_6, ˆa_7] = ˆvalue_1;

      if (ˆa_5 === 1 && ˆa_6 === 2 && ˆa_7 === ni) {
        ni;
        break ˆmatch_2;
      }
    }

    {
      shrub;
      break ˆmatch_2;
    }
  }
}
{
  ˆmatch_9: {
    const ˆvalue_8 = shrub;

    if (ˆvalue_8 !== undefined && ˆvalue_8 !== null) {
      const {
        \\"foo\\": ˆp_10,
        \\"ni\\": ˆp_11
      } = ˆvalue_8;

      if (ˆp_10 === 4 && ˆp_11 === ni) {
        spam + ni;
        break ˆmatch_9;
      }
    }

    if (ˆvalue_8 === 123) {
      spam;
      break ˆmatch_9;
    }
  }
}
{
  ˆmatch_13: {
    const ˆvalue_12 = shrub;

    if (ˆvalue_12 !== undefined && ˆvalue_12 !== null) {
      const {
        \\"foo\\": ˆp_14,
        \\"ni\\": ˆp_15
      } = ˆvalue_12;

      if (ˆp_14 === 4) {
        if (ˆp_15 !== undefined && ˆp_15 !== null) {
          const {
            \\"na\\": ˆp_16,
            \\"nu\\": ˆp_17
          } = ˆp_15;

          if (ˆp_16 === na && ˆp_17 === nu) {
            spam + ni;
            break ˆmatch_13;
          }
        }
      }
    }

    if (ˆvalue_12 !== undefined && ˆvalue_12 !== null && typeof ˆvalue_12[Symbol.iterator] === \\"function\\") {
      const [ˆa_18, ˆa_19, ˆa_20] = ˆvalue_12;

      if (ˆa_18 === 1 && ˆa_19 === 2) {
        if (ˆa_20 !== undefined && ˆa_20 !== null && typeof ˆa_20[Symbol.iterator] === \\"function\\") {
          const [ˆa_21] = ˆa_20;

          if (ˆa_21 === ni) {
            ni;
            break ˆmatch_13;
          }
        }
      }
    }

    if (ˆvalue_12 === 123) {
      spam;
      break ˆmatch_13;
    }
  }
}
Object.assign(module.exports, {});"
`;
